### PR TITLE
Fix detection for out of date files

### LINKED ISSUE #9359: Some buildernames does not support caching
I noticed that e.g. HTML builder can cache output and does not generate anything if there's no change:

```
$ time make -j16 html
sphinx-build -b "html" -d _build/doctrees  -j auto  . "_build/html"
Running Sphinx v4.0.2
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
no targets are out of date.
build succeeded.

The HTML pages are in _build/html.

real	0m0.686s
user	0m0.610s
sys	0m0.076s
```

while e.g. manpages generation always writes:

```
$ time make -j16 man
sphinx-build -b "man" -d _build/doctrees  -j auto  . "_build/man"
Running Sphinx v4.0.2
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [man]: all manpages
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
writing... gcc.1 { gcc-command-options/description } done
build succeeded.

The manual pages are in _build/man.

real	0m5.142s
user	0m5.009s
sys	0m0.133s
```

As seen, `looking for now-outdated files... none found` is printed but output files are re-generated.

### PR DESCRIPTION
**Pull request by marxin**
Fix detection when targets are out of date.

Fixes #9359.

### CODE REVIEW

**Comment by tk0miya** [comment]
Thank you for your proposal. It seems some tests are failed. Could you check them please?

**Comment by marxin** [comment]
Well, I am not fully sure why the tests fail. I removed 2x `shared_result` which fixed some tests.
Apart from that I see:
```
app = <SphinxTestApp buildername='texinfo'>, status = <_io.StringIO object at 0x7ffff0876a60>, warning = <_io.StringIO object at 0x7ffff0876af0>

    @pytest.mark.sphinx('texinfo', testroot='markup-rubric')
    def test_texinfo_rubric(app, status, warning):
        app.build()
        print(app)
    
>       output = (app.outdir / 'python.texi').read_text()

tests/test_build_texinfo.py:74: 
```

but when I run only a single test, it works fine `pytest tests/test_build_texinfo.py`. So it's likely about sharing `SphinxTestApp` among tests? Can you please help me?

**Comment by marxin** [comment]
I've got a tentative patch that works. But as said, I'm not much experienced with Sphinx.

**Comment by marxin** [comment]
@tk0miya May I please ping this pull request?

**Comment by marxin** [comment]
May I please ping this pull request?

**Comment by marxin** [comment]
Rebased this pull request. Is it something @AA-Turner can take a look at, please?

File: tests/test_build_latex.py [hunk 1]
```diff
@@ -1604,7 +1604,7 @@ def test_latex_container(app, status, warning):
 
 @pytest.mark.sphinx('latex', testroot='reST-code-role')
 def test_latex_code_role(app):
-    app.build()
```
Comment (AA-Turner) [inline tests/test_build_latex.py:1607]: Please use kwargs here

File: tests/test_build_manpage.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ def test_man_pages_empty_description(app, status, warning):
 @pytest.mark.sphinx('man', testroot='basic',
                     confoverrides={'man_make_section_directory': True})
 def test_man_make_section_directory(app, status, warning):
-    app.build()
+    app.build(True)
```
Comment (AA-Turner) [inline tests/test_build_manpage.py]: Please use kwargs here

File: tests/test_build_texinfo.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def test_texinfo(app, status, warning):
 
 @pytest.mark.sphinx('texinfo', testroot='markup-rubric')
 def test_texinfo_rubric(app, status, warning):
-    app.build()
+    app.build(True)
```
Comment (AA-Turner) [inline tests/test_build_texinfo.py]: Please use kwargs here

**Review by AA-Turner — CHANGES_REQUESTED** [review]

**Review by AA-Turner — APPROVED** [review]

**Comment by AA-Turner** [comment]
Please add a CHANGES entry

A

**Comment by marxin** [comment]
> Please add a CHANGES entry

Done.

File: CHANGES [hunk 4]
```diff
@@ -16,6 +16,8 @@ Features added
 * #10840: One can cross-reference including an option value: ``:option:`--module=foobar```.
   Patch by Martin Liska.
 * #10881: autosectionlabel: Record the generated section label to the debug log.
+* #9360: Fix caching for now-outdates files for some builders when there is not change
+  in source files. Patch by Martin Liska.
```
Comment (AA-Turner) [inline CHANGES]:
```suggestion
* #9360: Fix caching for now-outdated files for some builders (e.g. manpage)
  when there is no change in source files. Patch by Martin Liska.
```

**Comment by AA-Turner** [comment]
Thanks @marxin!

A

**Comment by marxin** [comment]
I thank you!

**Pull request merged**